### PR TITLE
Refactor changeset state into GraphQL API

### PR DESCRIPTION
**Pull request by eseliger**
We've had quite some effort to do in the frontend to map all the internal states on a changeset onto something that makes sense for the user. Now that I needed to touch it _again_ to add a 'retrying' state, I figured it's about time to do the refactor: Introduce the changeset state as a first-class citizen in the GraphQL API, which also allows to use it directly for filtering.

Closes https://github.com/sourcegraph/sourcegraph/issues/16009

### CODE REVIEW

File: client/web/src/enterprise/campaigns/close/CampaignCloseChangesetsList.tsx [hunk 1]
```diff
@@ -69,9 +64,10 @@ export const CampaignCloseChangesetsList: React.FunctionComponent<Props> = ({
     const queryChangesetsConnection = useCallback(
         (args: FilteredConnectionQueryArguments) =>
             queryChangesets({
-                externalState: ChangesetExternalState.OPEN,
-                publicationState: ChangesetPublicationState.PUBLISHED,
-                reconcilerState: null,
+                // TODO: This doesn't account for draft changesets. Ideally, this would
+                // use the delta API and apply an empty campaign spec, but then changesets
+                // would currently be lost.
```
Comment (eseliger) [inline client/web/src/enterprise/campaigns/close/CampaignCloseChangesetsList.tsx:69]: No change to before

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 2]
```diff
@@ -1648,17 +1648,9 @@ type Campaign implements Node {
...
-        reconcilerState: [ChangesetReconcilerState!]
-        """
-        Only include changesets with the given publication state.
-        """
-        publicationState: ChangesetPublicationState
-        """
-        Only include changesets with the given external state.
-        """
-        externalState: ChangesetExternalState
+        state: ChangesetState
```
Comment (LawnGnome) [inline cmd/frontend/graphqlbackend/schema.graphql:1653]: I doubt anyone external has built anything on top of this yet, but now we're not beta, this is the kind of change we need to be more aware of in terms of breaking backward compatibility. I think this is fine for now, but we need to start tuning our review heuristics to pick this kind of thing up.

File: client/web/src/enterprise/campaigns/detail/changesets/ExternalChangesetInfoCell.tsx [hunk 3]
```diff
@@ -28,12 +26,7 @@ export const ExternalChangesetInfoCell: React.FunctionComponent<ExternalChangese
         <div className="m-0">
             <h3 className="m-0 d-block d-md-inline">
                 <LinkOrSpan
-                    /* Deleted changesets most likely don't exist on the codehost anymore and would return 404 pages */
-                    to={
-                        node.externalURL && node.externalState !== ChangesetExternalState.DELETED
-                            ? node.externalURL.url
-                            : undefined
-                    }
+                    to={node.externalURL?.url ?? undefined}
```
Comment (LawnGnome) [inline client/web/src/enterprise/campaigns/detail/changesets/ExternalChangesetInfoCell.tsx:29]: Ah, null coalesce. PHP's greatest gift to the world of language design?

**Review by LawnGnome — APPROVED** [review]
My reaction to this is a bit like my reaction to the related #16860: there's a pedantic part of me that doesn't really like conflating states together, but I can also see how much simpler this makes the UI (check that diffstat!), and I appreciate that this only really permeates down to the resolvers, below which we keep the existing distinctions.

So I think I'm good with this, overall. It manages the combinatorial explosion of states we were starting to have to grapple within the UI, and doesn't weaken our internal representation of the disjoint states. Nice job!

File: cmd/frontend/graphqlbackend/schema.graphql (hunk 2 above)
Comment (eseliger) [reply cmd/frontend/graphqlbackend/schema.graphql:1653]: Yeah, makes sense. I wasn't 100% sure about this either but also wanted to be sure that no place referencing these fields is left in the code-base as we don't want to use those fields anymore, if we don't need to.
I'll re-add those fields and deprecate them for now.

**Pull request merged**